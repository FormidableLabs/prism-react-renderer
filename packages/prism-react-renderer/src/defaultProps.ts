import Prism from "./vendor/prism"
import theme from "./themes/duotoneDark"

const defaultProps = {
  Prism,
  theme,
}
export default defaultProps
