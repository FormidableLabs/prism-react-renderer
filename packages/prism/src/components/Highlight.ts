import { Component } from "react"
import normalizeTokens from "../utils/normalizeTokens"
import type { ThemeDict } from "../utils/themeToDict"
import themeToDict from "../utils/themeToDict"
import type { Token as PrismToken } from "prismjs"
// import { loadLanguages } from "prismjs/components"
import type {
  Language,
  Token,
  LineInputProps,
  LineOutputProps,
  TokenInputProps,
  TokenOutputProps,
  RenderProps,
  PrismGrammar,
  PrismLib,
  PrismTheme,
  EnvConfig,
} from "../types"

type Props = {
  Prism: PrismLib
  theme?: PrismTheme
  language: Language
  code: string
  children: (props: RenderProps) => JSX.Element
}

// loadLanguages(["jsx"])

class Highlight extends Component<Props, unknown> {
  prevTheme?: PrismTheme
  prevLanguage?: Language
  themeDict?: ThemeDict
  getThemeDict = (props: Props): ThemeDict | void => {
    if (
      this.themeDict !== undefined &&
      props.theme === this.prevTheme &&
      props.language === this.prevLanguage
    ) {
      return this.themeDict
    }

    this.prevTheme = props.theme
    this.prevLanguage = props.language
    const themeDict = props.theme
      ? themeToDict(props.theme, props.language)
      : undefined
    return (this.themeDict = themeDict)
  }
  getLineProps = ({
    className,
    style,
    ...rest
  }: LineInputProps): LineOutputProps => {
    const output: LineOutputProps = {
      ...rest,
      className: "token-line",
      style: undefined,
      key: undefined,
    }
    const themeDict = this.getThemeDict(this.props)

    if (themeDict !== undefined) {
      output.style = themeDict.plain
    }

    if (style !== undefined) {
      output.style =
        output.style !== undefined ? { ...output.style, ...style } : style
    }

    if (className) output.className += ` ${className}`
    return output
  }
  getStyleForToken = ({ types, empty }: Token) => {
    const typesSize = types.length
    const themeDict = this.getThemeDict(this.props)

    if (themeDict === undefined) {
      return undefined
    } else if (typesSize === 1 && types[0] === "plain") {
      return empty
        ? {
            display: "inline-block",
          }
        : undefined
    } else if (typesSize === 1 && !empty) {
      return themeDict[types[0]]
    }

    const baseStyle = empty
      ? {
          display: "inline-block",
        }
      : {}
    const typeStyles = types.map((type: string) => themeDict[type])
    return Object.assign(baseStyle, ...typeStyles)
  }
  getTokenProps = ({
    key,
    className,
    style,
    token,
    ...rest
  }: TokenInputProps): TokenOutputProps => {
    const output: TokenOutputProps = {
      ...rest,
      className: `token ${token.types.join(" ")}`,
      children: token.content,
      style: this.getStyleForToken(token),
      key: undefined,
    }

    if (style !== undefined) {
      output.style =
        output.style !== undefined ? { ...output.style, ...style } : style
    }

    if (key !== undefined) output.key = key
    if (className) output.className += ` ${className}`
    return output
  }
  tokenize = (
    Prism: PrismLib,
    code: string,
    grammar: PrismGrammar,
    language: Language
  ): Array<PrismToken | string> => {
    const env: EnvConfig = {
      code,
      grammar,
      language,
      tokens: [],
    }
    Prism.hooks.run("before-tokenize", env)
    const tokens = (env.tokens = Prism.tokenize(env.code, env.grammar))
    Prism.hooks.run("after-tokenize", env)
    return tokens
  }

  render() {
    const { Prism, language, code, children } = this.props
    const themeDict = this.getThemeDict(this.props)
    const grammar = Prism.languages[language]
    const mixedTokens =
      grammar !== undefined
        ? this.tokenize(Prism, code, grammar, language)
        : [code]
    const tokens = normalizeTokens(mixedTokens)
    return children({
      tokens,
      className: `prism-code language-${language}`,
      style: themeDict !== undefined ? themeDict.root : {},
      getLineProps: this.getLineProps,
      getTokenProps: this.getTokenProps,
    })
  }
}

export default Highlight
