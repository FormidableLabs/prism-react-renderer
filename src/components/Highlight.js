// @flow

import React, { Component, type Node } from "react";
import normalizeTokens from "../utils/normalizeTokens";
import themeToDict, { type ThemeDict } from "../utils/themeToDict";

import type {
  Language,
  Token,
  LineInputProps,
  LineOutputProps,
  TokenInputProps,
  TokenOutputProps,
  RenderProps,
  PrismLib,
  PrismTheme
} from "../types";

type Props = {
  Prism: PrismLib,
  theme?: PrismTheme,
  language: Language,
  code: string,
  children: (props: RenderProps) => Node
};

class Highlight extends Component<Props, *> {
  prevTheme: PrismTheme | void;
  prevLanguage: Language | void;
  themeDict: ThemeDict | void;

  getThemeDict = (props: Props) => {
    if (
      this.themeDict !== undefined &&
      props.theme === this.prevTheme &&
      props.language === this.prevLanguage
    ) {
      return this.themeDict;
    }

    this.prevTheme = props.theme;
    this.prevLanguage = props.language;

    const themeDict = props.theme
      ? themeToDict(props.theme, props.language)
      : undefined;
    return (this.themeDict = themeDict);
  };

  getLineProps = ({
    key,
    className,
    style,
    line,
    ...rest
  }: LineInputProps): LineOutputProps => {
    const output: LineOutputProps = {
      ...rest,
      className: "token-line",
      style: undefined,
      key: undefined
    };

    const themeDict = this.getThemeDict(this.props);
    if (themeDict !== undefined) {
      output.style = themeDict.plain;
    }

    if (style !== undefined) {
      output.style =
        output.style !== undefined ? { ...output.style, ...style } : style;
    }

    if (key !== undefined) output.key = key;
    if (className) output.className += ` ${className}`;

    return output;
  };

  getStyleForToken = ({ types, empty }: Token) => {
    const typesSize = types.length;
    const themeDict = this.getThemeDict(this.props);

    if (themeDict === undefined) {
      return undefined;
    } else if (typesSize === 1 && types[0] === "plain") {
      return empty ? { display: "inline-block" } : undefined;
    } else if (typesSize === 1 && !empty) {
      return themeDict[types[0]];
    }

    const baseStyle = empty ? { display: "inline-block" } : {};
    // $FlowFixMe
    const typeStyles = types.map(type => themeDict[type]);
    return Object.assign(baseStyle, ...typeStyles);
  };

  getTokenProps = ({
    key,
    className,
    style,
    token,
    ...rest
  }: TokenInputProps): TokenOutputProps => {
    const output: TokenOutputProps = {
      ...rest,
      className: `token ${token.types.join(" ")}`,
      children: token.content,
      style: this.getStyleForToken(token),
      key: undefined
    };

    if (style !== undefined) {
      output.style =
        output.style !== undefined ? { ...output.style, ...style } : style;
    }

    if (key !== undefined) output.key = key;
    if (className) output.className += ` ${className}`;

    return output;
  };

  render() {
    const { Prism, language, code, children } = this.props;

    const themeDict = this.getThemeDict(this.props);

    const grammar = Prism.languages[language];
    const mixedTokens =
      grammar !== undefined ? Prism.tokenize(code, grammar, language) : [code];
    const tokens = normalizeTokens(mixedTokens);

    return children({
      tokens,
      className: `prism-code language-${language}`,
      style: themeDict !== undefined ? themeDict.root : {},
      getLineProps: this.getLineProps,
      getTokenProps: this.getTokenProps
    });
  }
}

export default Highlight;
