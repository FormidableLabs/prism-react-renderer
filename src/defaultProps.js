// @flow

import Prism from "./vendor/prism";
import theme from "./themes/duotoneDark";

import type { PrismLib } from "./types";

const defaultProps = {
  // $FlowFixMe
  Prism: (Prism: PrismLib),
  theme
};

export default defaultProps;
