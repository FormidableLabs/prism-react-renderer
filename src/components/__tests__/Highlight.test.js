import React from "react";
import { render, cleanup } from "react-testing-library";

import Highlight from "../Highlight";
import defaultProps from "../../defaultProps";

const exampleCode = `
(function someDemo() {
  var test = "Hello World!";
  console.log(test);
})();

return () => <App />;
`.trim();

describe("<Highlight />", () => {
  afterEach(cleanup);

  describe("snapshots", () => {
    it("renders correctly", () => {
      const { container } = render(
        <Highlight {...defaultProps} code={exampleCode} language="jsx">
          {({ className, style, tokens, getLineProps, getTokenProps }) => (
            <pre className={className} style={style}>
              {tokens.map((line, i) => (
                <div {...getLineProps({ line, key: i })}>
                  {line.map((token, key) => (
                    <span {...getTokenProps({ token, key })} />
                  ))}
                </div>
              ))}
            </pre>
          )}
        </Highlight>
      );

      expect(container).toMatchSnapshot();
    });

    it("renders unsupported languages correctly", () => {
      const { container } = render(
        <Highlight
          {...defaultProps}
          code={exampleCode}
          language="abcdefghijklmnop"
        >
          {({ className, style, tokens, getLineProps, getTokenProps }) => (
            <pre className={className} style={style}>
              {tokens.map((line, i) => (
                <div {...getLineProps({ line, key: i })}>
                  {line.map((token, key) => (
                    <span {...getTokenProps({ token, key })} />
                  ))}
                </div>
              ))}
            </pre>
          )}
        </Highlight>
      );

      expect(container).toMatchSnapshot();
    });

    it("renders without style props when no theme is passed", () => {
      const { container } = render(
        <Highlight
          {...defaultProps}
          theme={undefined}
          code={exampleCode}
          language="jsx"
        >
          {({ className, style, tokens, getLineProps, getTokenProps }) => (
            <pre className={className} style={style}>
              {tokens.map((line, i) => (
                <div {...getLineProps({ line, key: i })}>
                  {line.map((token, key) => (
                    <span {...getTokenProps({ token, key })} />
                  ))}
                </div>
              ))}
            </pre>
          )}
        </Highlight>
      );

      expect(container.innerHTML.includes("style")).toBeFalsy();
    });
  });

  describe("getLineProps", () => {
    it("transforms lineProps inputs correctly", () => {
      const input = {
        key: "line-1",
        style: { cssProp: "test" },
        className: "line-class",
        line: [{ types: ["punctuation"], content: "!" }],
        restPropsTest: true
      };

      render(
        <Highlight {...defaultProps} code={exampleCode} language="jsx">
          {({ getLineProps }) => {
            const output = getLineProps(input);

            expect(output).toEqual({
              key: "line-1",
              style: {
                cssProp: "test",
                backgroundColor: null,
                color: expect.any(String)
              },
              className: "token-line line-class",
              restPropsTest: true
            });

            return null;
          }}
        </Highlight>
      );
    });
  });

  describe("getTokenProps", () => {
    it("transforms tokenProps inputs correctly", () => {
      const input = {
        key: "token-1",
        style: { cssProp: "test" },
        className: "token-class",
        token: { types: ["punctuation"], content: "!" },
        restPropsTest: true
      };

      render(
        <Highlight {...defaultProps} code={exampleCode} language="jsx">
          {({ getTokenProps }) => {
            const output = getTokenProps(input);

            expect(output).toEqual({
              key: "token-1",
              style: { cssProp: "test", color: expect.any(String) },
              className: "token punctuation token-class",
              restPropsTest: true,
              children: "!"
            });

            return null;
          }}
        </Highlight>
      );
    });

    it("transforms constructor token style correctly", () => {
      // From https://github.com/FormidableLabs/prism-react-renderer/issues/11
      render(
        <Highlight {...defaultProps} code={"open Common;"} language="reason">
          {({ tokens, getTokenProps }) => {
            const line = tokens[0];
            const token = line[2];
            const output = getTokenProps({ token, key: 2 });

            expect(typeof output.style).not.toBe("function");

            return null;
          }}
        </Highlight>
      );
    });
  });
});
