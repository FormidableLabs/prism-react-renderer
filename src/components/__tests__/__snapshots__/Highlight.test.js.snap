// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Highlight /> snapshots renders correctly 1`] = `
<div>
  <pre
    class="prism-code language-jsx"
    style="background-color: rgb(42, 39, 52); color: rgb(154, 134, 253);"
  >
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        (
      </span>
      <span
        class="token keyword"
        style="color: rgb(255, 204, 153);"
      >
        function
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(154, 134, 253);"
      >
        someDemo
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        )
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        {
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      >
          
      </span>
      <span
        class="token keyword"
        style="color: rgb(255, 204, 153);"
      >
        var
      </span>
      <span
        class="token plain"
      >
         test 
      </span>
      <span
        class="token operator"
        style="color: rgb(224, 145, 66);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token string"
        style="color: rgb(255, 204, 153);"
      >
        "Hello World!"
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        ;
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      >
          console
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        .
      </span>
      <span
        class="token function"
        style="color: rgb(154, 134, 253);"
      >
        log
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        (
      </span>
      <span
        class="token plain"
      >
        test
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        ;
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      />
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        }
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        ;
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
        style="display: inline-block;"
      />
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      />
      <span
        class="token keyword"
        style="color: rgb(255, 204, 153);"
      >
        return
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        )
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token operator"
        style="color: rgb(224, 145, 66);"
      >
        =&gt;
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token tag punctuation"
        style="color: rgb(108, 103, 131);"
      >
        &lt;
      </span>
      <span
        class="token tag class-name"
        style="color: rgb(224, 145, 66);"
      >
        App
      </span>
      <span
        class="token tag"
        style="color: rgb(224, 145, 66);"
      >
         
      </span>
      <span
        class="token tag punctuation"
        style="color: rgb(108, 103, 131);"
      >
        /&gt;
      </span>
      <span
        class="token punctuation"
        style="color: rgb(108, 103, 131);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`<Highlight /> snapshots renders unsupported languages correctly 1`] = `
<div>
  <pre
    class="prism-code language-abcdefghijklmnop"
    style="background-color: rgb(42, 39, 52); color: rgb(154, 134, 253);"
  >
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      >
        (function someDemo() {
      </span>
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      >
          var test = "Hello World!";
      </span>
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      >
          console.log(test);
      </span>
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      >
        })();
      </span>
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
        style="display: inline-block;"
      />
    </div>
    <div
      class="token-line"
      style="color: rgb(154, 134, 253);"
    >
      <span
        class="token plain"
      >
        return () =&gt; &lt;App /&gt;;
      </span>
    </div>
  </pre>
</div>
`;
